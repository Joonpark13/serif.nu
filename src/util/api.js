/* eslint-disable import/prefer-default-export */
export function searchURL(query) {
  return `${process.env.API_URL}/search?term_id=4720&search_query=${query}`;
}

export function sectionsURL(termId, schoolAbbv, subjectAbbv, courseAbbv) {
  return `${process.env.API_URL}/sections?term_id=${termId}&school_abbv=${schoolAbbv}&subject_abbv=${subjectAbbv}&course_abbv=${courseAbbv}`;
}
