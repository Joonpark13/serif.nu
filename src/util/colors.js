export const northwesternPurple = '#4E2A84';
export const northwesternBrightOrange = '#FFC520';
export const northwesternRichBlack10 = '#D8D6D6';
