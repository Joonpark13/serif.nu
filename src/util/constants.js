/* eslint-disable import/prefer-default-export */
export const CURRENT_TERM = '4730';
