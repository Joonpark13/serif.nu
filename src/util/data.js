import termsData from 'data/terms.json';

/* eslint-disable import/prefer-default-export */
/* istanbul ignore next */
export const currentTerm = termsData.reduce((highestTerm, term) => {
  if (term.id > highestTerm.id) {
    return term;
  }
  return highestTerm;
});
/* eslint-enable import/prefer-default-export */
