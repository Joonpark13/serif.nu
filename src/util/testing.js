/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import * as reactRedux from 'react-redux';
import { shallow } from 'enzyme';
import * as useSelector from './use-selector';

jest.mock('./use-selector');

// Helper function that returns a function which creates enzyme wrappers with default props
export function wrapperCreator(Component, defaultProps = {}) {
  let classes;
  return props => shallow(<Component classes={classes} {...defaultProps} {...props} />);
}

/* istanbul ignore next */
export function mockUseSelector(...args) {
  useSelector.default.mockReset();
  args.forEach(arg => useSelector.default.mockReturnValueOnce(arg));

  return useSelector.default;
}

/* istanbul ignore next */
export function mockUseDispatch() {
  const dispatchSpy = jest.fn();
  reactRedux.useDispatch.mockReturnValue(dispatchSpy);
  return dispatchSpy;
}
