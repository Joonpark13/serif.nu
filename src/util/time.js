export function formatMinute(minute) {
  if (minute < 10) return `0${minute}`;
  return `${minute}`;
}

export function formatTime(time, includeAmPm = true) {
  if (time.hour === 0) return `12:${formatMinute(time.minute)}${includeAmPm ? ' AM' : ''}`;
  if (time.hour < 12) return `${time.hour}:${formatMinute(time.minute)}${includeAmPm ? ' AM' : ''}`;
  if (time.hour === 12) return `12:${formatMinute(time.minute)}${includeAmPm ? ' PM' : ''}`;
  return `${time.hour - 12}:${formatMinute(time.minute)}${includeAmPm ? ' PM' : ''}`;
}

export function getFormattedClassEvent(event) {
  if (event.dow === 'TBA' || event.start === 'TBA' || event.end === 'TBA') {
    return 'TBA';
  }
  return `${formatTime(event.start, false)} - ${formatTime(event.end, false)}`;
}

export function getFormattedClassSchedule(event) {
  if (event.dow === 'TBA' || event.start === 'TBA' || event.end === 'TBA') {
    return 'TBA';
  }
  const dowStr = event.dow.join('');
  return `${dowStr} ${formatTime(event.start, true)} - ${formatTime(event.end, true)}`;
}

export function getDurationInHours(event) {
  const hoursDiff = event.end.hour - event.start.hour;
  const minutesDiff = event.end.minute - event.start.minute;
  return hoursDiff + minutesDiff / 60;
}

export function isBefore(time1, time2, orEqual = false) {
  if (time1.hour < time2.hour) {
    return true;
  }
  if (time1.hour === time2.hour) {
    if (orEqual) return time1.minute <= time2.minute;
    return time1.minute < time2.minute;
  }
  return false;
}

export function overlaps(event1, event2) {
  if (event1.dow !== event2.dow) {
    return false;
  }
  return !(isBefore(event1.end, event2.start) || isBefore(event2.end, event1.start));
}
