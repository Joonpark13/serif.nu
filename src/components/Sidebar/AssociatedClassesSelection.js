import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from '@material-ui/core/Typography';
import { getFormattedClassSchedule } from 'util/time';
import SidebarHeader from './SidebarHeader';

export const styles = {
  sectionsRoot: {
    padding: 16,
  },
  title: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  divider: {
    marginTop: 5,
  },
};

function AssociatedClassesSelection({
  currentCourseName,
  currentSectionNumber,
  associatedClasses,
  back,
  addSectionWithAssociatedClass,
  classes,
}) {
  return (
    <div className={classes.sectionsRoot}>
      <SidebarHeader title={`${currentCourseName}-${currentSectionNumber}`} back={back} />

      <List>
        {associatedClasses.map(associatedClass => (
          <ListItem
            key={JSON.stringify(associatedClass.schedule)}
            button
            onClick={() => addSectionWithAssociatedClass(associatedClass)}
          >
            <ListItemText>
              <Typography variant="h6">
                {associatedClass.type}
              </Typography>

              <Typography>
                {getFormattedClassSchedule(associatedClass.schedule)}
              </Typography>
              <Typography>{associatedClass.schedule.location}</Typography>
            </ListItemText>
          </ListItem>
        ))}
      </List>
    </div>
  );
}

AssociatedClassesSelection.propTypes = {
  currentCourseName: PropTypes.string.isRequired,
  currentSectionNumber: PropTypes.string.isRequired,
  associatedClasses: PropTypes.arrayOf(PropTypes.object).isRequired,
  back: PropTypes.func.isRequired,
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  addSectionWithAssociatedClass: PropTypes.func.isRequired,
};

export { AssociatedClassesSelection as UnstyledAssociatedClassesSelection };
export default withStyles(styles)(AssociatedClassesSelection);
