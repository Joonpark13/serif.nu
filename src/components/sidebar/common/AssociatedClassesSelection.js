import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/styles';
import { List, ListItem, ListItemText, Typography } from '@material-ui/core';
import { getFormattedClassSchedule, isUnscheduled } from 'util/time';
import SidebarHeader from '../SidebarHeader';

const useStyles = makeStyles({
  sectionsRoot: {
    padding: 16,
  },
  title: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  divider: {
    marginTop: 5,
  },
});

export default function AssociatedClassesSelection({
  currentCourseName,
  currentSectionNumber,
  associatedClasses,
  back,
  addSectionWithAssociatedClass,
  associatedClassHover,
  associatedClassHoverOff,
}) {
  const classes = useStyles();
  return (
    <div className={classes.sectionsRoot}>
      <SidebarHeader title={`${currentCourseName}-${currentSectionNumber}`} back={back} />

      <List>
        {associatedClasses.map((associatedClass) => {
          const associatedClassIsUnscheduled = isUnscheduled(associatedClass.schedule);
          return (
            <ListItem
              key={JSON.stringify(associatedClass.schedule)}
              button
              onClick={() => addSectionWithAssociatedClass(associatedClass)}
              onMouseEnter={() => associatedClassHover(associatedClass)}
              onMouseLeave={associatedClassHoverOff}
              disabled={associatedClassIsUnscheduled}
            >
              <ListItemText>
                <Typography variant="h6">
                  {associatedClass.type}
                </Typography>

                <Typography color={associatedClassIsUnscheduled ? 'error' : undefined}>
                  {getFormattedClassSchedule(associatedClass.schedule)}
                </Typography>
                <Typography>{associatedClass.schedule.location}</Typography>
              </ListItemText>
            </ListItem>
          );
        })}
      </List>
    </div>
  );
}

AssociatedClassesSelection.propTypes = {
  currentCourseName: PropTypes.string.isRequired,
  currentSectionNumber: PropTypes.string.isRequired,
  associatedClasses: PropTypes.arrayOf(PropTypes.object).isRequired,
  back: PropTypes.func.isRequired,
  addSectionWithAssociatedClass: PropTypes.func.isRequired,
  associatedClassHover: PropTypes.func.isRequired,
  associatedClassHoverOff: PropTypes.func.isRequired,
};
