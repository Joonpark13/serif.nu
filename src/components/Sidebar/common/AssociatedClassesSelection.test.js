import ListItem from '@material-ui/core/ListItem';
import { wrapperCreator } from 'util/testing';
import * as timeUtils from 'util/time';
import { UnstyledAssociatedClassesSelection, styles } from './AssociatedClassesSelection';

jest.mock('util/time');

describe('AssociatedClassesSelection', () => {
  const associatedClass = { schedule: { location: 'somewhere' } };
  const defaultProps = {
    currentCourseName: 'Introduction to Something',
    currentSectionNumber: '21',
    associatedClasses: [associatedClass],
    back: () => {},
    addSectionWithAssociatedClass: () => {},
    associatedClassHover: () => {},
    associatedClassHoverOff: () => {},
  };
  const getWrapper = wrapperCreator(UnstyledAssociatedClassesSelection, defaultProps, styles);

  const formattedSchedule = 'MWF 10 - 12ish';
  timeUtils.getFormattedClassSchedule.mockReturnValue(formattedSchedule);

  it('renders correctly', () => {
    const wrapper = getWrapper();

    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('adds section with associated class when clicked', () => {
    const addSectionWithAssociatedClassMock = jest.fn();
    const wrapper = getWrapper({
      addSectionWithAssociatedClass: addSectionWithAssociatedClassMock,
    });
    wrapper.find(ListItem).simulate('click');

    expect(addSectionWithAssociatedClassMock).toHaveBeenCalledWith(associatedClass);
  });

  it('calls associatedClassHover when clicked', () => {
    const associatedClassHoverMock = jest.fn();
    const wrapper = getWrapper({
      associatedClassHover: associatedClassHoverMock,
    });
    wrapper.find(ListItem).simulate('mouseEnter');

    expect(associatedClassHoverMock).toHaveBeenCalledWith(associatedClass);
  });
});
