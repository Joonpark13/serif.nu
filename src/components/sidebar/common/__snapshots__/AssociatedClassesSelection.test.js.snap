// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssociatedClassesSelection renders correctly 1`] = `
<div
  className=""
>
  <ForwardRef(WithStyles)
    back={[Function]}
    title="Introduction to Something-21"
  />
  <ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      button={true}
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <ForwardRef(WithStyles)>
        <ForwardRef(WithStyles)
          variant="h6"
        />
        <ForwardRef(WithStyles)>
          MWF 10 - 12ish
        </ForwardRef(WithStyles)>
        <ForwardRef(WithStyles)>
          somewhere
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
    </ForwardRef(WithStyles)>
  </ForwardRef(WithStyles)>
</div>
`;
